### DISCUSSION TITLE
Catching error messages to show meaningful explanation of failures

### DISCUSSION THREAD

**Discussion by kryaksy** [perspective-dev/perspective#2480] [Q&A]
## Support Question

I've encountered an issue where various CSV conversion errors are being logged to the console, which makes it challenging to capture and display meaningful messages to the user. These errors seem to abort the update process unexpectedly.

Attached is a screenshot illustrating the errors:
<img width="1373" alt="Screenshot 2023-12-20 at 14 20 56 copy" src="https://github.com/finos/perspective/assets/33466807/e04837eb-2205-4841-beec-674acd4fa7b7">

I am looking for a way to properly capture these errors so that I can provide a meaningful and user-friendly response instead of these raw messages. The current behavior is not conducive to a good user experience, as the console logs are not visible to the end-users.

Could you advise on best practices for capturing these errors within the code? 

Ideally, I would like to intercept these errors before they hit the console, and then log them or display a custom error message to the user. Any pointers or examples would be greatly appreciated.

**Comment by texodus** [comment]
I don't understand - these aren't user errors, they are errors in your usage of the API (sending invalid data to `update()`). What is the use case for reporting these?

(edit - I've moved this to a discussion below (rather than Issue), GitHub's UI does not make this clear IMO since it uses the same color for "closed" vs "moved")

**Reply by kryaksy** [reply]
When users upload large and complex CSV files, like in my case, finding errors during upload can be tricky. It's really important to have clear error messages. These messages should help users figure out what's going wrong, even if we can't point out the exact spot of the error in the file. Sure, we could use basic external checks, but using the advanced error-handling in Perspective’s API is a much better way.

**Reply by texodus** [reply]
> When users upload large and complex CSV files

Again, this only happens _during calls to `update()`_, not generally for all errors. Initial table construction will throw the complete error, including line and column number:

<img width="982" alt="Screenshot 2024-01-03 at 1 50 25 PM" src="https://github.com/finos/perspective/assets/60666/64bc05a5-74ab-4c71-bdab-276a816ebf18">

Why are you sending _user provided (dirty) data_ to `update()`? There is AFAIK no use case at all for this.

**Reply by tomjakubowski** [reply]
> Why are you sending user provided (dirty) data to update()? There is AFAIK no use case at all for this.

You raise a good point. CSV is a finnicky format, underspecified, and with lots of CSV data out there only loosely conforming to what specifications do exist.

How would a user of the perspective API validate CSV strings to ensure that they are compatible with Perspective's CSV parser?

**Reply by kryaksy** [reply]
> Initial table construction will throw the complete error, including line and column number:

@texodus Good to know. However, I've just experimented this and it's not throwing data type conversion errors like:
<img width="966" alt="image" src="https://github.com/finos/perspective/assets/33466807/96b3e79a-cce0-4d3e-9665-10cbcdee5117">

> Why are you sending user provided (dirty) data to update()? There is AFAIK no use case at all for this.

I am not sure which case I should sample other than the need to understand why the perspective API is unfavorable towards the CSV file and to guide the user in correcting that mistake? 

> How would a user of the perspective API validate CSV strings to ensure that they are compatible with Perspective's CSV parser?

Thanks @tomjakubowski. This question might be a more constructive one I think, pointing the issue better.

**Reply by texodus** [reply]
> However, I've just experimented this and it's not throwing data type conversion errors like

Please post repros of what you tried, not just screenshots of errors. This works just fine:

<img width="828" alt="Screenshot 2024-01-08 at 7 03 38 PM" src="https://github.com/finos/perspective/assets/60666/2aa2de18-9809-417e-914e-6aada5e2a446">

**Reply by kryaksy** [reply]
It seems I made a mistake in my previous explanation – apologies for the confusion. As you pointed out, the initial construction does indeed throw errors. However, it's clear now that conversion errors are not a concern when creating a table.

The actual scenario involves updating an existing table with a poorly formatted CSV file, rather than creating a new one. Therefore, my focus returns to handling errors that may arise during the table.update process. If you need, I can prepare a repro for this case.

**Reply by texodus** [reply]
> The actual scenario involves updating an existing table with a poorly formatted CSV file, rather than creating a new one.

Please explain the _actual scenario_, then.

**Reply by kryaksy** [reply]
We're having trouble importing data into our predefined database tables with CSV files. The data types in the CSV don't match what our tables are set up for. Does this cover the _actual scenario_ you were asking about? If not, please guide me on how to explain it better.

**Reply by texodus** [reply]
No, that's about as clear as mud. The `table()` constructor does exactly what you are asking for - it will throw validation errors, and you can use `table.schema()` to check what column types Perspective inferred if you want to pass this _validated_ data along to another predefined database. Trying to do this by creating an empty Perspective table and updating it is not the intended use case for `update()`, as I've said several times now this is the wrong method to use for user input validation - this method is for _updating_ a Perspective Table with more data from a server that has already performed validation, not from a user.

**Reply by kryaksy** [reply]
> you can use table.schema() to check what column types Perspective inferred if you want to pass this validated data along to another predefined database

Thanks for the suggestion. It's definitely a more suitable and acceptable method compared to our current approach. Ideally, though, we would prefer to intercept the messages that Perspective is already managing. While it might seem like I'm insisting on something trivial, it's really about gaining access to a feature you already have.

**Comment by texodus** [comment]
> While it might seem like I'm insisting on something trivial, it's really about gaining access to a feature you already have.

Adding error handling to `update()` is the _farthest_ thing from trivial.
